crates/dht/src/time.rs

```diff
@@ -0,0 +1,1034 @@
+//!
+//! As an example, consider a factor of 9. This means that full slices are 2^9 = 512 times the [UNIT_TIME].
+//! With a unit time of 15 minutes, that means full slices are 512 * 15 minutes = 7680 minutes = 128 hours.
+//! So every 5 days, a new full slice is created.
+//! The partial slices reserve at least 2^8 + 2^7 ... 2^0 = 519 times the [UNIT_TIME], so 519 * 15 minutes
+//! = 7785 minutes = 129.75 hours. That gives roughly 5 days of recent time.
+//!
+//! A lower factor allocates less recent time and more slices to be stored but is more granular when
+//! comparing time slices with another peer. A higher factor allocates more recent time and fewer
+//! slices to be stored but is less granular when comparing time slices with another peer.
+
+use crate::constant::UNIT_TIME;
+use kitsune2_api::{DynOpStore, K2Error, K2Result, OpId, Timestamp};
+use std::time::Duration;
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone, PartialEq))]
+pub struct PartitionedTime {
+    /// The timestamp at which the time partitioning starts.
+    ///
+    /// This must be a constant among peers who need to compute the same time slices.
+    origin_timestamp: Timestamp,
+    /// The factor used to determine the size of the time slices.
+    ///
+    /// The size of a time slice is 2^factor * [UNIT_TIME].
+    /// Partial slices are created for "recent time" and are created in decreasing sizes from
+    /// 2^factor * [UNIT_TIME] down to 2^0 * [UNIT_TIME].
+    factor: u8,
+    /// The number of full time slices that have been stored.
+    ///
+    /// These full slices are always of size 2^factor * [UNIT_TIME].
+    full_slices: u64,
+    /// The partial slices, wish hashes stored in memory.
+    ///
+    /// Because these change every [UNIT_TIME], they are stored in memory.
+    partial_slices: Vec<PartialSlice>,
+    /// The minimum amount of time that must be reserved for recent time.
+    ///
+    /// The algorithm is free to allocate up to 2x this value as recent time. This allows for the
+    /// algorithm to create more slices that fill more of the recent time window.
+    min_recent_time: Duration,
+    /// The timestamp at which the next update is required.
+    ///
+    /// After this time, there will be an excess amount of time that could be allocated into a
+    /// partial slice. The data structure is still usable but will get behind if
+    /// [PartitionedTime::update] is not called.
+    /// It is idempotent to call [PartitionedTime::update] more often than required, but it is not
+    /// efficient.
+    next_update_at: Timestamp,
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone, PartialEq))]
+pub struct PartialSlice {
+    /// The start timestamp of the time slice.
+    ///
+    /// This timestamp is included in the time slice.
+    start: Timestamp,
+    /// Size is used in the formula 2^size * [UNIT_TIME].
+    ///
+    /// That means a 0 size translates to [UNIT_TIME], and a 1 size translates to 2*[UNIT_TIME].
+    size: u8,
```

Comment:
I think I'm actually fine with this as-is, but consider if it would be clearer to test and reason about if you split the concerns of the time slicing and the hash maintenance:

```rust
pub struct TimeSliceLayout {
    pub full_slice_count: u64,
    pub partial_slice_list: Vec<(Timestamp, u8)>,
}

pub fn calc_current_time_slices(origin_timestamp: Timestamp, now: Timestamp) -> TimeSliceLayout {
    // ...
}

impl PartitionedTime {
    pub async fn update(
        &mut self,
        store: DynOpStore,
        layout: TimeSliceLayout,
    ) -> K2Result<()> {
        // ...
    }
}
```

Reply:
Trying to picture what this would look like. I think the update would need to return a diff or something like that. That's simple enough for the full slices but it's tricky for the partial slices where the hashes are kept in memory and we'd have to keep both old and new around to be able to figure out which changed. 

Definitely possible, and although I was thinking about these being one concern, this could be split. Do you think it's worthwhile?

Reply:
That's true, I might just be shifting the complexity around with this.

Although to clarify my suggestion, the layout function wouldn't have the hash in it, just like my example code, so I feel like you could iterate the new layout list and pluck the old one out by start time if it exists, doing the existing comparison if it exists and either re-hashing or not.

Part of the time it took me here to achieve enough understanding to review was in figuring out the interplay between the figuring out where the time slices are bounded, and then how that is used to generate the hashes. I *feel* like I might be able to get up to speed faster if I could set aside the time slice logic and focus only on the hash generation.

But I don't want to make things worse, so I'll rely on your judgment from here! : )

Reply:
Right, I have a better idea of what I should change to answer this question now. Yes I think this is valuable to do now because `update` is a rather large function and it is doing two things. I would like to keep all the logic contained inside this type but I think it can be split into two functions that do different things. I'll have a shot at this now :)

Reply:
I think this is better and easier to digest, what do you think? Does it go far enough?

[e9729db7](https://github.com/holochain/kitsune2/pull/20/commits/e9729db7c1063a390357a38e190483e08746ff47)

Reply:
Yes, that is easier to follow, thanks!